### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by hconrad**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    "log"
+  ],
+  "author": "Hans Conrad",
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/hconrad/winston-transport-blob-storage/issues"
+  },
+  "homepage": "https://github.com/hconrad/winston-transport-blob-storage#readme",
+  "dependencies": {
+    "@azure/storage-blob": "^10.2.0-preview",
```
Comment (ChuckLangford) [inline package.json]: Any need to consider concreting these versions?

File: lib/transport-blob-storage.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  ServiceURL,
+  ContainerURL,
+  AppendBlobURL,
+  Aborter,
+  SharedKeyCredential,
+  StorageURL,
+} = require('@azure/storage-blob');
+
+function checkStringValue(value) {
+  if(!value || typeof value !== 'string' || value.length === 0) {
```
Comment (ChuckLangford) [inline lib/transport-blob-storage.js]: Linting

File: README.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# Winston Transport for Azure Blob Storage
+
+This simple transport allows log statements to be directly uploaded to Blob Storage
```
Comment (ChuckLangford) [inline README.md:3]: Realize you're not done but suggest adding a solid writeup on how to use this code. Examples and such.

File: lib/transport-blob-storage.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    throw new Error('Invalid value supplied');
+  }
+}
+
+module.exports = class AzureBlobStorageTransport extends Transport {
+  constructor(opts) {
+    super(opts);
```
Comment (ChuckLangford) [inline lib/transport-blob-storage.js:20]: Do you need to check opts? Does it exist? Is it an object?

File: lib/transport-blob-storage.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    const {
+      accountName, accountKey, storageUrl, containerName,
```
Comment (ChuckLangford) [inline lib/transport-blob-storage.js:25]: How does the code respond if these values are not in opts? (admit I haven't read the testing code yet).

File: lib/transport-blob-storage.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    } = opts;
+    Object.values(opts).forEach((val) => checkStringValue(val));
```
Comment (ChuckLangford) [inline lib/transport-blob-storage.js]: Can't decide. Should this happen after the call to super(opts)?

File: lib/transport-blob-storage.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  checkError(info, appendBlob) {
+    return (error) => {
+      if (error.statusCode !== 404) {
+        throw error;
+      }
+      if (!error.message || !error.message) {
+        throw error;
+      }
+      const errorMessage = error.message.toLowerCase();
+      if (errorMessage.includes('container')) {
```
Comment (ChuckLangford) [inline lib/transport-blob-storage.js:43]: So both of these "errors" are a 404?

File: lib/transport-blob-storage.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      } else {
+        throw error;
+      }
+    };
+  }
+
+  logMessage(appendBlob, info) {
+    const message = `${info.message}\n`;
+    const logSize = message.length;
+    return appendBlob.appendBlock(Aborter.none, info.message, logSize).catch(this.checkError(info, appendBlob));
```
Comment (ChuckLangford) [inline lib/transport-blob-storage.js]: What does this return? Why does it need to return?

File: lib/transport-blob-storage.test.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+function createMockAppendBlob() {
+    return {
+        create: jest.fn().mockResolvedValue({}),
+        appendBlock: jest.fn().mockResolvedValue({}),
+    }
+}
+
+describe('transport-blob-storage', () => {
+    test('constructor param validation', () => {
```
Comment (ChuckLangford) [inline lib/transport-blob-storage.test.js]: So I think this partially answers my question. I wonder if you need to put more validation in place around the parameter.

File: lib/transport-blob-storage.test.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        const badProps = createTransportProps({accountName: undefined});
+        expect(() => new Transport(badProps)).toThrow('Invalid value supplied');
+        const goodProps = createTransportProps({});
+        const transport = new Transport(goodProps);
+        expect(transport.containerUrl).toBeDefined();
+    });
+    test('checkError exception paths', () => {
```
Comment (ChuckLangford) [inline lib/transport-blob-storage.test.js]: seems like this doesn't cover all the code paths

File: package.json (hunk 1 above)
Comment (hconrad) [reply package.json]: yup good call
```suggestion
    "@azure/storage-blob": "10.2.0-preview",
```

File: lib/transport-blob-storage.js (hunk 2 above)
Comment (hconrad) [reply lib/transport-blob-storage.js]:
```suggestion
  if (!value || typeof value !== 'string' || value.length === 0) {
```

Comment (hconrad) [reply lib/transport-blob-storage.js]: yeah also still have to add the .eslintrc file

File: README.md (hunk 3 above)
Comment (hconrad) [reply README.md:3]: Agreed and yes haven't gotten to that yet

File: lib/transport-blob-storage.js (hunk 5 above)
Comment (hconrad) [reply lib/transport-blob-storage.js:25]: as mentioned below I should be able to move the validation code above the assignment. which will also make sure opt exists.

File: lib/transport-blob-storage.js (hunk 6 above)
Comment (hconrad) [reply lib/transport-blob-storage.js]: I think it should be above

File: lib/transport-blob-storage.js (hunk 7 above)
Comment (hconrad) [reply lib/transport-blob-storage.js:43]: that's correct.

File: lib/transport-blob-storage.test.js (hunk 10 above)
Comment (hconrad) [reply lib/transport-blob-storage.test.js]: covers around 92% of them. I'm sure I can further expand on the test cases :)

File: lib/transport-blob-storage.js (hunk 2 above)
Comment (hconrad) [reply lib/transport-blob-storage.js]: Added .eslintr cand fixed

File: lib/transport-blob-storage.js (hunk 4 above)
Comment (hconrad) [reply lib/transport-blob-storage.js:20]: yeah good call added

**Pull request merged**